jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartInputStreamParser.java

```diff
@@ -0,0 +1,794 @@
+        if (_err != null)
+        {
+            _err.addSuppressed(new Throwable());
+            if (_err instanceof IOException)
+                throw (IOException)_err;
+            if (_err instanceof IllegalStateException)
+                throw (IllegalStateException)_err;
+            throw new IllegalStateException(_err);
+        }
+    }
+
+    /**
+     * Parse, if necessary, the multipart stream.
+     *
+     */
+    protected void parse ()
+    {
+        //have we already parsed the input?
+        if (_parts != null || _err != null)
+            return;
+        try
+        {
+
+            //initialize
+            _parts = new MultiMap<>();
+
+            //if its not a multipart request, don't parse it
+            if (_contentType == null || !_contentType.startsWith("multipart/form-data"))
+                return;
+
+
+            //sort out the location to which to write the files
+            if (_config.getLocation() == null)
+                _tmpDir = _contextTmpDir;
+            else if ("".equals(_config.getLocation()))
+                _tmpDir = _contextTmpDir;
+            else
+            {
+                File f = new File (_config.getLocation());
+                if (f.isAbsolute())
+                    _tmpDir = f;
+                else
+                    _tmpDir = new File (_contextTmpDir, _config.getLocation());
+            }
+
+            if (!_tmpDir.exists())
+                _tmpDir.mkdirs();
+
+            String contentTypeBoundary = "";
+            int bstart = _contentType.indexOf("boundary=");
+            if (bstart >= 0)
+            {
+                int bend = _contentType.indexOf(";", bstart);
+                bend = (bend < 0? _contentType.length(): bend);
+                contentTypeBoundary = QuotedStringTokenizer.unquote(value(_contentType.substring(bstart,bend)).trim());
+            }
+
+
+            Handler handler = new Handler();
+            MultiPartParser parser = new MultiPartParser(handler,contentTypeBoundary);            
+
+
+            // Create a buffer to store data from stream //
+            byte[] data = new byte[_bufferSize];
+            int len=0;
+
+            /* keep running total of size of bytes read from input 
+             * and throw an exception if exceeds MultipartConfigElement._maxRequestSize */
+            long total = 0; 
+
+            while(true) 
+            {
+
+                len = _in.read(data);
+
+                if(len > 0) 
+                {
+                    total+=len;
+                    if(_config.getMaxRequestSize() > 0 && total > _config.getMaxRequestSize())
+                    {
+                        _err = new IllegalStateException ("Request exceeds maxRequestSize ("+_config.getMaxRequestSize()+")");
+                        return;
+                    }
+
+                    ByteBuffer buffer = BufferUtil.toBuffer(data);
+                    buffer.limit(len);
+                    parser.parse(buffer, false);
```

Comment:
what if the parser does not read all of the data from the buffer?

Reply:
what if the parser returns false?

Reply:
The MultiPartParser will fully consume the buffer unless it throws an exception or an exception is set in the handler. The handler is set to only return true on `messageComplete()` or if an exception is set in _err. So I will need to change line 579 to `if(parser.parse(buffer, false))   break;` so it correctly exits on exceptions from the handler.

Reply:
So fully consuming and only returning true on messageComplete() are assumptions about the parser and handler.  You should either code in a way that does not make those assumptions... or at least code to check the assumptions are true and throw an exception if they are not.

Reply:
Why is it an assumption if the handler is implemented directly below on line 620.
But I could put a check to make sure the buffer is fully consumed if it is safer.

Reply:
It's still an assumption, just a lot more locally made one.  So it's not so bad as the handler is in the same class.... but in future it might get factored out into its own class and move further away.   So checking that the call to parse returns true (AND the parser is in end state) is a simple and cheap way to record/check that assumption.     Remember that for the rest of us, we deal with lots of other parsers, not all of them fully consuming and many of them will return true for partially parsed content... so we need to be able to switch between code bases and quickly understand what kind of parser we are dealing with.  Having the assumptions checked in code will help with that.